src/Framework/AspNetCoreAnalyzers/src/Analyzers/WebApplicationBuilder/PublicPartialProgramClassAnalyzer.cs

```diff
@@ -0,0 +1,45 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.AspNetCore.Analyzers;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class PublicPartialProgramClassAnalyzer : DiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DiagnosticDescriptors.PublicPartialProgramClassNotRequired);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+        context.RegisterSyntaxNodeAction(context =>
+        {
+            var syntaxNode = context.Node;
+            if (IsPublicPartialClassProgram(syntaxNode))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(
+                    DiagnosticDescriptors.PublicPartialProgramClassNotRequired,
+                    syntaxNode.GetLocation()));
+            }
+        }, SyntaxKind.ClassDeclaration);
+    }
+
+    private static bool IsPublicPartialClassProgram(SyntaxNode syntaxNode)
+    {
+        return syntaxNode is ClassDeclarationSyntax { Modifiers: { } modifiers } classDeclaration
+            && (classDeclaration.Members == null || classDeclaration.Members.Count == 0) // Skip non-empty declarations
+            && modifiers is { Count: > 1 }
+            && modifiers.Any(SyntaxKind.PublicKeyword)
+            && modifiers.Any(SyntaxKind.PartialKeyword)
+            && classDeclaration is { Identifier.ValueText: "Program" }
+            && classDeclaration.Parent is CompilationUnitSyntax parentNode
```

Comment:
I'm wondering whether the order matters here; thinking about the fastest way we can exclude misses... not benchmarked, but I'd expect the `is CompilationUnitSyntax`, as a raw type test, to be very fast *and* exclude almost everything, and the `Identifier.ValueText` test to mop up almost everything else, so perhaps *only then* (once we know we're dealing with `[...] class Program [...]` test the modifiers & members?

Reply:
Yep, I agree with this categorization. You only need to sweat the details once you've evaluated you're in a top-level `Program` definition. I'll update the order of checks accordingly.